massa-event-cache/src/event_cache.rs

```diff
@@ -0,0 +1,454 @@
+use std::cmp::max;
+// std
+use nom::AsBytes;
+use std::path::Path;
+// third-party
+use crate::ser_deser::{
+    SCOutputEventDeserializer, SCOutputEventDeserializerArgs, SCOutputEventSerializer,
+};
+use massa_models::execution::EventFilter;
+use massa_models::output_event::SCOutputEvent;
+use massa_serialization::{DeserializeError, Deserializer, Serializer};
+use rocksdb::{DBIterator, IteratorMode, WriteBatch, DB};
+use tracing::debug;
+
+const OPEN_ERROR: &str = "critical: rocksdb open operation failed";
+const CRUD_ERROR: &str = "critical: rocksdb crud operation failed";
+const EVENT_DESER_ERROR: &str = "critical: event deserialization failed";
+
+/// Module key formatting macro
+#[macro_export]
+macro_rules! event_key {
+    ($event_slot:expr) => {
+        [&$event_slot.to_bytes()[..], &[MODULE_IDENT]].concat()
+    };
+}
+
+pub(crate) struct EventCache {
+    /// RocksDB database
+    db: DB,
+    /// How many entries are in the db. Count is initialized at creation time by iterating
+    /// over all the entries in the db then it is maintained in memory
+    entry_count: usize,
+    /// Maximum number of entries we want to keep in the db.
+    /// When this maximum is reached `snip_amount` entries are removed
+    max_entry_count: usize,
+    /// How many entries are removed when `entry_count` reaches `max_entry_count`
+    snip_amount: usize,
+    /// Event serializer
+    event_ser: SCOutputEventSerializer,
+    /// Event deserializer
+    event_deser: SCOutputEventDeserializer,
+}
+
+impl EventCache {
+    /// Create a new EventCache
+    pub fn new(path: &Path, max_entry_count: usize, snip_amount: usize, thread_count: u8) -> Self {
+        let db = DB::open_default(path).expect(OPEN_ERROR);
+        let entry_count = db.iterator(IteratorMode::Start).count();
+
+        Self {
+            db,
+            entry_count,
+            max_entry_count,
+            snip_amount,
+            event_ser: SCOutputEventSerializer::new(),
+            event_deser: SCOutputEventDeserializer::new(SCOutputEventDeserializerArgs {
+                thread_count,
+            }),
+        }
+    }
+
+    #[allow(dead_code)]
+    /// Insert a new event in the cache
+    pub fn insert(&mut self, event: SCOutputEvent) {
+        if self.entry_count >= self.max_entry_count {
+            self.snip(None);
+        }
+
+        let event_key = {
+            let mut event_key = event.context.slot.to_bytes_key().to_vec();
+            event_key.extend(event.context.index_in_slot.to_be_bytes());
+            event_key
+        };
+        let mut event_buffer = Vec::new();
+        self.event_ser.serialize(&event, &mut event_buffer).unwrap();
+
+        let mut batch = WriteBatch::default();
+        batch.put(event_key, event_buffer);
+        self.db.write(batch).expect(CRUD_ERROR);
+        self.entry_count = self.entry_count.saturating_add(1);
```

Comment:
note: this assumes that events are always added, never overwritten. This is a correct assumption but should be written here as a comment

Reply:
Added a comment with this assumption